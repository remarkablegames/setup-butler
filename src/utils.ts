import os from 'os';
import path from 'path';

const Arch = {
  arm64: 'amd64',
  arm: '386',
  x32: '386',
  x64: 'amd64',
} as const;

/**
 * Gets the operating system CPU architecture.
 *
 * @see {@link https://nodejs.org/api/os.html#os_os_arch}
 *
 * @returns - Return value in [386, amd64]
 */
function getArch() {
  const arch = Arch[os.arch() as keyof typeof Arch];
  if (arch) {
    return arch;
  }
  throw new Error(`Unsupported arch: ${os.arch()}`);
}

enum Platform {
  darwin = 'darwin',
  linux = 'linux',
  win32 = 'windows',
}

/**
 * Gets a string identifying the operating system platform.
 *
 * @see {@link https://nodejs.org/api/os.html#os_os_platform}
 *
 * @returns - Return value in [darwin, linux, windows]
 */
function getPlatform() {
  const platform = Platform[os.platform() as keyof typeof Platform];
  if (platform) {
    return platform;
  }
  throw new Error(`Unsupported platform: ${os.platform()}`);
}

/**
 * Gets download URL.
 *
 * @see {@link https://itch.io/docs/butler/installing.html}
 *
 * @param version - CLI version
 * @returns - Download URL
 */
export function getDownloadUrl(version: string) {
  return `https://broth.itch.zone/butler/${getPlatform()}-${getArch()}/${version}/archive/default`;
}

/**
 * Gets CLI path.
 *
 * @param directory - Directory
 * @param name - CLI name
 * @returns - Binary path
 */
export function getBinaryPath(directory: string, name: string) {
  return path.join(
    directory,
    name + (getPlatform() === Platform.win32 ? '.exe' : ''),
  );
}
