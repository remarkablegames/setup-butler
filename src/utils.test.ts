import os from 'os';

import { getBinaryPath, getDownloadUrl } from './utils';

jest.mock('os');
const mockedOs = jest.mocked(os);

const platforms: NodeJS.Platform[] = ['darwin', 'linux', 'win32'];
const architectures = ['arm', 'arm64', 'x32', 'x64'] as NodeJS.Architecture[];

describe('getDownloadUrl', () => {
  const version = '1.2.3';

  const table = platforms.reduce(
    (testSuites, os) => [
      ...testSuites,
      ...architectures.map(
        (arch) => [os, arch] as [NodeJS.Platform, NodeJS.Architecture],
      ),
    ],
    [] as [NodeJS.Platform, NodeJS.Architecture][],
  );

  beforeEach(() => {
    jest.resetAllMocks();
  });

  describe.each(table)('when platform is %p and arch is %p', (os, arch) => {
    beforeEach(() => {
      mockedOs.platform.mockReturnValue(os);
      mockedOs.arch.mockReturnValue(arch);
    });

    it('returns download object', () => {
      expect(getDownloadUrl(version)).toMatchSnapshot();
    });
  });

  describe('error', () => {
    const arch = 'ia32';

    it(`throws error when architecture is "${arch}"`, () => {
      mockedOs.arch.mockReturnValue(arch);
      mockedOs.platform.mockReturnValue('linux');
      expect(() => {
        getDownloadUrl(version);
      }).toThrow(`Unsupported arch: ${arch}`);
    });

    const platform = 'aix';

    it(`throws error when platform is "${platform}"`, () => {
      mockedOs.arch.mockReturnValue('x64');
      mockedOs.platform.mockReturnValue(platform);
      expect(() => {
        getDownloadUrl(version);
      }).toThrow(`Unsupported platform: ${platform}`);
    });
  });
});

describe('getBinaryPath', () => {
  describe.each(platforms)('when platform is %p', (os) => {
    beforeEach(() => {
      jest.resetAllMocks();
      mockedOs.platform.mockReturnValue(os);
    });

    it('returns CLI path', () => {
      const directory = 'directory';
      const name = 'name';
      expect(getBinaryPath(directory, name)).toMatchSnapshot();
    });
  });
});
