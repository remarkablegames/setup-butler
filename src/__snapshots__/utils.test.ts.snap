// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBinaryPath when platform is "darwin" returns CLI path 1`] = `"directory/name"`;

exports[`getBinaryPath when platform is "linux" returns CLI path 1`] = `"directory/name"`;

exports[`getBinaryPath when platform is "win32" returns CLI path 1`] = `"directory/name.exe"`;

exports[`getDownloadUrl when platform is "darwin" and arch is "arm" returns download object 1`] = `"http://broth.itch.ovh/butler/darwin-386/1.2.3/archive/default"`;

exports[`getDownloadUrl when platform is "darwin" and arch is "arm64" returns download object 1`] = `"http://broth.itch.ovh/butler/darwin-amd64/1.2.3/archive/default"`;

exports[`getDownloadUrl when platform is "darwin" and arch is "x32" returns download object 1`] = `"http://broth.itch.ovh/butler/darwin-386/1.2.3/archive/default"`;

exports[`getDownloadUrl when platform is "darwin" and arch is "x64" returns download object 1`] = `"http://broth.itch.ovh/butler/darwin-amd64/1.2.3/archive/default"`;

exports[`getDownloadUrl when platform is "linux" and arch is "arm" returns download object 1`] = `"http://broth.itch.ovh/butler/linux-386/1.2.3/archive/default"`;

exports[`getDownloadUrl when platform is "linux" and arch is "arm64" returns download object 1`] = `"http://broth.itch.ovh/butler/linux-amd64/1.2.3/archive/default"`;

exports[`getDownloadUrl when platform is "linux" and arch is "x32" returns download object 1`] = `"http://broth.itch.ovh/butler/linux-386/1.2.3/archive/default"`;

exports[`getDownloadUrl when platform is "linux" and arch is "x64" returns download object 1`] = `"http://broth.itch.ovh/butler/linux-amd64/1.2.3/archive/default"`;

exports[`getDownloadUrl when platform is "win32" and arch is "arm" returns download object 1`] = `"http://broth.itch.ovh/butler/windows-386/1.2.3/archive/default"`;

exports[`getDownloadUrl when platform is "win32" and arch is "arm64" returns download object 1`] = `"http://broth.itch.ovh/butler/windows-amd64/1.2.3/archive/default"`;

exports[`getDownloadUrl when platform is "win32" and arch is "x32" returns download object 1`] = `"http://broth.itch.ovh/butler/windows-386/1.2.3/archive/default"`;

exports[`getDownloadUrl when platform is "win32" and arch is "x64" returns download object 1`] = `"http://broth.itch.ovh/butler/windows-amd64/1.2.3/archive/default"`;
